packages/devtools_app/lib/src/timeline/flutter/timeline_flame_chart.dart

```diff
@@ -337,6 +349,240 @@ class _FullTimelineFlameChartState
+        rowHeight;
+  }
+
+  double _calculateHorizontalGuidelineY(TimelineEvent event) {
+    final spacerRowsBeforeEvent = _spacerRowsBeforeEvent(event);
+    return spacerRowsBeforeEvent * sectionSpacing +
+        (chartNodesByEvent[event].row.index - spacerRowsBeforeEvent) *
+            rowHeightWithPadding +
+        rowHeight / 2;
+  }
+}
+
+class AsyncGuidelinePainter extends CustomPainter {
+  AsyncGuidelinePainter({
+    @required this.zoom,
+    @required this.constraints,
+    @required this.verticalScrollOffset,
+    @required this.horizontalScrollOffset,
+    @required this.verticalGuidelines,
+    @required this.horizontalGuidelines,
+  });
+
+  final double zoom;
+
+  final BoxConstraints constraints;
+
+  final double verticalScrollOffset;
+
+  final double horizontalScrollOffset;
+
+  List<VerticalLineSegment> verticalGuidelines;
+
+  List<HorizontalLineSegment> horizontalGuidelines;
+
+  @override
+  void paint(Canvas canvas, Size size) {
+    final Rect visible = Rect.fromLTWH(
+      horizontalScrollOffset,
+      verticalScrollOffset,
+      constraints.maxWidth,
+      constraints.maxHeight,
+    );
+
+    final firstVerticalGuidelineIndex = lowerBound(
+      verticalGuidelines,
+      VerticalLineSegment(visible.topLeft, visible.bottomLeft),
+    );
+    final firstHorizontalGuidelineIndex = lowerBound(
+      horizontalGuidelines,
+      HorizontalLineSegment(visible.topLeft, visible.topRight),
+    );
+
+    if (firstVerticalGuidelineIndex != -1) {
+      _paintGuidelines(
+        canvas,
+        visible,
+        verticalGuidelines,
+        firstVerticalGuidelineIndex,
+      );
+    }
+
+    if (firstHorizontalGuidelineIndex != -1) {
+      _paintGuidelines(
+        canvas,
+        visible,
+        horizontalGuidelines,
+        firstHorizontalGuidelineIndex,
+      );
+    }
+  }
+
+  void _paintGuidelines(
+    Canvas canvas,
+    Rect visible,
+    List<LineSegment> guidelines,
+    int firstLineIndex,
+  ) {
+    for (int i = firstLineIndex; i < guidelines.length; i++) {
+      final line = guidelines[i];
+
+      // We are out of range on the cross axis.
+      if (!line.crossAxisIntersects(visible)) break;
+
+      // Only paint lines that intersect [visible] along both axes.
+      if (line.intersects(visible)) {
+        canvas.drawLine(
+          Offset(
+            (line.start.dx - horizontalScrollOffset)
+                .clamp(0.0, constraints.maxWidth),
+            (line.start.dy - verticalScrollOffset)
+                .clamp(0.0, constraints.maxHeight),
+          ),
+          Offset(
+            (line.end.dx - horizontalScrollOffset)
+                .clamp(0.0, constraints.maxWidth),
+            (line.end.dy - verticalScrollOffset)
+                .clamp(0.0, constraints.maxHeight),
+          ),
+          Paint()..color = treeGuidelineColor,
+        );
+      }
+    }
+  }
+
+  @override
+  bool shouldRepaint(AsyncGuidelinePainter oldDelegate) => true;
```

Comment:
does shouldRepaint have to be true? maybe add a TODO to return false when the delegate is the same?

Reply:
added todo